lib/JitEngineDevice.hpp

```diff
@@ -658,15 +608,59 @@ JitEngineDevice<ImplT>::compileAndRun(
   // used by any kernel that will be specialized
   auto JitModule = llvm::CloneModule(extractDeviceBitcode(KernelName, Kernel));
   // NOTE: There is potential oportunity here, to reduce some of the JIT costs
-  // further. We can have a specializeIR in which we do not do any RC/Grid/Block
+  // further. We can have a pruneIR in which we do not do any RC/Grid/Block
   // specializations. We only internalize symbols. Then we can use that IR
   // for all upcoming specializations of dynamic information.
   // There is a memory trade off in such case, We will need to have a peristent
   // in memory module, for every annotated kernel. If we have a case of 1000s of
   // kernels, this can be an issue
 
-  specializeIR(*JitModule, KernelName, Suffix, BlockDim, GridDim, RCIndices,
-               RCsVec.data(), NumRuntimeConstants);
+  auto *JitFunction = pruneIR(*JitModule, KernelName);
+
+  // Internalize others besides the kernel function.
+  internalizeModule(*JitModule, [&JitFunction](const GlobalValue &GV) {
+    // Do not internalize the kernel function.
+    if (&GV == JitFunction)
+      return true;
+
+    // Internalize everything else.
+    return false;
+  });
+
+  // Replace argument uses with runtime constants.
+  if (Config.ENV_PROTEUS_SPECIALIZE_ARGS)
+    // TODO: change NumRuntimeConstants to size_t at interface.
+    TransformArgumentSpecialization::transform(
+        *JitModule, *JitFunction, RCIndices,
+        ArrayRef<RuntimeConstant>{RCsVec.data(),
+                                  static_cast<size_t>(NumRuntimeConstants)});
+
+  // Replace uses of blockDim.* and gridDim.* with constants.
+  if (Config.ENV_PROTEUS_SPECIALIZE_DIMS)
+    setKernelDims(*JitModule, GridDim, BlockDim);
+
+  PROTEUS_DBG(Logger::logs("proteus")
+              << "=== JIT Module\n"
+              << *JitModule << "=== End of JIT Module\n");
+
+  JitFunction->setName(KernelName + Suffix);
+
+  if (Config.ENV_PROTEUS_SET_LAUNCH_BOUNDS)
+    setLaunchBoundsForKernel(*JitModule, *JitFunction,
+                             GridDim.x * GridDim.y * GridDim.z,
+                             BlockDim.x * BlockDim.y * BlockDim.z);
+
+  runCleanupPassPipeline(*JitModule);
+
+#if PROTEUS_ENABLE_DEBUG
+  Logger::logs("proteus") << "=== Final Module\n"
+                          << *JitModule << "=== End Final Module\n";
+  if (verifyModule(*JitModule, &errs()))
+    FATAL_ERROR("Broken module found, JIT compilation aborted!");
+  else
+    Logger::logs("proteus") << "Module verified!\n";
+#endif
```

Comment:
Re-instate `specializeIR` and put this code in its body. Get `Function *F` as needed from the `Module`